### ISSUE TITLE
Real-time collaborative editing with editor.js

### ISSUE THREAD

**Issue by Psyphant** [codex-team/editor.js#908]
Hey!

Did someone test to get editor.js work with SwellRT, Convergence Labs or other solutions to make the editing collaborative?

**Comment by oliver-dvorski** [comment]
You'd still need to write quite a bit of logic to make these two communicate, no?

**Comment by milu2003** [comment]
i need it too!

**Comment by marcoancona** [comment]
I tried to get real-time collaboration with Firebase Real-time Database but editor.js lacks fundamental APIs to get this to work nicely.

In particular, without a mechanism to render only some blocks rather than all, a mechanism to pass metadata to blocks or to make individual blocks read-only, it is very hard.

**Comment by milu2003** [comment]
> I tried to get real-time collaboration with Firebase Real-time Database but editor.js lacks fundamental APIs to get this to work nicely.
> 
> In particular, without a mechanism to render only some blocks rather than all, a mechanism to pass metadata to blocks or to make individual blocks read-only, it is very hard.

editorJS didn't think of this from the beginning of design. Multi person real-time collaboration is not only a problem of local rendering, such as how to undo multi person real-time collaboration?

How to compare the differences between the two versions. It is difficult to develop these functions using editorjs. I gave up

How to recover from one version to another is very difficult

**Comment by marcoancona** [comment]
My idea was to take advantage of the block nature of the underlying data and to prevent editing a block that is being edited by somebody else, by setting it read-only. Conflict resolution should be easier under the constraint that within the same block there should be no conflicts. Undo only for the changes within the block being edited. This should be possible now that blocks have a unique id. On the other hand, for this to work it should be possible for all other blocks to be updated without disrupting the editing experience of the block the user is working on.

**Comment by milu2003** [comment]
As for what you said about locking a block for editing, I think the user experience is very bad

At the beginning of my project, I used editor to develop it. Later, I didn't support collaboration well enough. Later, I used prosemirror to reconstruct the project. I found that prosemirror did a good job in this aspect. Editor has been abandoned by me.

You can also try tiptap, based on prosemirror：
https://www.tiptap.dev/examples/collaborative-editing#example

I think if real-time collaboration is an important feature of your product, then editorjs is not a better choice, because  the design didn't take this into account at the beginning.

**Comment by hata6502** [comment]
This issue may be related to... https://github.com/codex-team/editor.js/pull/1606

**Comment by marcoancona** [comment]
Tiptap seems interesting, thanks for the hint @milu2003. Although, I have to say, that I would prefer a block-based editor to build something closer to Notion/Confluence. And editor.js looks good out of the box.

@hata6502 indeed that would help

**Comment by milu2003** [comment]
Block building is not difficult. My current product is also in the form of block, similar to notion, but it's very easy for me to build with prosemirror, just a little more work. But the high degree of customization of prosemirror made it easier for me to build programs. What you said out of the box is both good and bad. Although it's easy to use, once you want more complex functions, I think your program will be greatly limited.

@marcoancona

**Comment by wmonecke** [comment]
@marcoancona @milu2003 I will see if [https://togetherjs.com/](https://togetherjs.com/) is a good solution for this problem.

**Comment by wmonecke** [comment]
Have you guys looked at this issue? #1155 Looks promising!

**Comment by SavoInfostream** [comment]
Any progress update for this feature?

**Comment by neSpecc** [comment]
> Any progress update for this feature?

Yes, we've designed and started implementing it. Here is [the first part](https://github.com/codex-team/editor.js/pull/1801). 

You can support us and boost development speed via [donation](http://opencollective.com/editorjs) on OpenCollective.

**Comment by SavoInfostream** [comment]
Donated.
Keep it going :)
